x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/runtimefields/RuntimeFieldsFeatureSetUsage.java

```diff
@@ -0,0 +1,272 @@
+
+import org.elasticsearch.Version;
+import org.elasticsearch.cluster.metadata.IndexMetadata;
+import org.elasticsearch.cluster.metadata.MappingMetadata;
+import org.elasticsearch.common.io.stream.StreamInput;
+import org.elasticsearch.common.io.stream.StreamOutput;
+import org.elasticsearch.common.io.stream.Writeable;
+import org.elasticsearch.common.xcontent.ToXContentObject;
+import org.elasticsearch.common.xcontent.XContentBuilder;
+import org.elasticsearch.xpack.core.XPackFeatureSet;
+import org.elasticsearch.xpack.core.XPackField;
+
+import java.io.IOException;
+import java.util.Arrays;
+import java.util.Comparator;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Set;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+public class RuntimeFieldsFeatureSetUsage extends XPackFeatureSet.Usage {
+
+    public static RuntimeFieldsFeatureSetUsage fromMetadata(Iterable<IndexMetadata> metadata) {
+        Map<String, RuntimeFieldStats> fieldTypes = new HashMap<>();
+        for (IndexMetadata indexMetadata : metadata) {
+            if (indexMetadata.isSystem()) {
+                // Don't include system indices in statistics about mappings, we care about the user's indices.
+                continue;
+            }
+            Set<String> indexFieldTypes = new HashSet<>();
+            MappingMetadata mappingMetadata = indexMetadata.mapping();
+            if (mappingMetadata != null) {
+                Object runtimeObject = mappingMetadata.getSourceAsMap().get("runtime");
+                if (runtimeObject instanceof Map) {
+                    @SuppressWarnings("unchecked")
+                    Map<String, Object> runtimeMappings = (Map<String, Object>) runtimeObject;
+                    for (Object runtimeFieldMappingObject : runtimeMappings.values()) {
+                        if (runtimeFieldMappingObject instanceof Map) {
+                            @SuppressWarnings("unchecked")
+                            Map<String, Object> runtimeFieldMapping = (Map<String, Object>) runtimeFieldMappingObject;
+                            Object typeObject = runtimeFieldMapping.get("type");
+                            if (typeObject != null) {
+                                String type = typeObject.toString();
+                                RuntimeFieldStats stats = fieldTypes.computeIfAbsent(type, RuntimeFieldStats::new);
+                                stats.count++;
+                                if (indexFieldTypes.add(type)) {
+                                    stats.indexCount++;
+                                }
+                                Object scriptObject = runtimeFieldMapping.get("script");
+                                if (scriptObject == null) {
+                                    stats.scriptLessCount++;
+                                } else if (scriptObject instanceof Map) {
```

Comment:
If script isn't an `instanceof Map` can it be a `String`? Would we get the short form here?

Reply:
It is my understanding that regardless of how it was submitted, the script always gets outputted in the long form. I will double check though. That is also why I am using the Script class in the unit tests